File: lengthOfLongestSubstring.java
//Leetcode3: https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
class Solution {
    public int lengthOfLongestSubstring(String s) {
        var set = new HashSet<Character>();

        int max =0;
        int start =0;
        int end =0;

        while(end < s.length()){
            if(!set.contains(s.charAt(end))){
                if(end - start + 1 > max){
                    max = end - start + 1;
                }
                set.add(s.charAt(end));
                end++;
            }
            else {
                set.remove(s.charAt(start));
                start++;
            }
        }

        return max;
    }
}

File: LeetCode/SlidingWindow/lengthOfLongestSubstring.ts
/**
 https://leetcode.com/problems/longest-substring-without-repeating-characters/

 */
function lengthOfLongestSubstring(s: string): number {
  let left = 0;
  let longestSubstringSet = new Set();
  let longest = 0;

  for (let right = 0; right < s.length; right++) {
    while (longestSubstringSet.has(s[right])) {
      longestSubstringSet.delete(s[left]);
      left++;
    }

    longestSubstringSet.add(s[right]);
    longest = Math.max(longest, right - left + 1);
  }
  //   console.log(longestSubstringSet);

  return longest;
}

console.log("3. Longest substring without repeating char.");
console.log(lengthOfLongestSubstring("abcabcbb"));
console.log(lengthOfLongestSubstring("bbbb"));
console.log(lengthOfLongestSubstring("pwwkew"));
9;
